File: katas/6 kyu/which-are-in/solution.py
# which-are-in
import codewars_test as test
# -------------------------------- MY SOLUTION ------------------------------- #
def in_array(array1, array2):
  r = []
  for x in array2:
    for y in array1:
      if y in x:
        r.append(y)
  return sorted(list(set(r)))
# ------------------------------ CLEVER SOLUTION ----------------------------- #
def in_array(a1, a2):
  return sorted({sub for sub in a1 if any(sub in s for s in a2)})
# ----------------------------------- TEST ----------------------------------- #
@test.describe("in_array")
def tests():
    def testing(array1, array2, expect):
        actual = in_array(array1, array2)
        test.assert_equals(actual, expect)
    @test.it("Basic tests")
    def basics():
        a1 = ["live", "arp", "strong"] 
        a2 = ["lively", "alive", "harp", "sharp", "armstrong"]
        r = ['arp', 'live', 'strong']
        testing(a1, a2, r)          
        a1 = ["arp", "mice", "bull"] 
        a2 = ["lively", "alive", "harp", "sharp", "armstrong"]
        r = ['arp']
        testing(a1, a2, r)    

File: katas/6 kyu/which-are-in/solution.php
<!-- which-are-in -->
<?php
use PHPUnit\Framework\TestCase;
/* ------------------------------- MY SOLUTION ------------------------------ */
function inArray($array1, $array2) {
  $arr = array();
  foreach ($array2 as $x){
    foreach ($array1 as $y){
      if (strpos($x , $y) !== false && in_array($y , $arr) == false){
        array_push($arr , $y);
      }
    }
  }
  sort($arr);
  return $arr;
}
/* ----------------------------- CLEVER SOLUTION ---------------------------- */
function inArray($a1, $a2) {
  $r = array_filter($a1, function($v) use ($a2) {
    foreach ($a2 as $v2) {
      if (strpos($v2, $v) !== false) return true;
    }
    return false;
  });
  sort($r);
  return $r;
}
/* ---------------------------------- TEST ---------------------------------- */
class WhichAreInTestCases extends TestCase
{
    private function revTest($actual, $expected) {
        $this->assertSame($expected, $actual);
    }
    public function testBasics() {
        $a2 = ["lively", "alive", "harp", "sharp", "armstrong"];
        $a1 = ["arp", "live", "strong"];
        $this->revTest(inArray($a1, $a2), ["arp", "live", "strong"]);
        $a1 = ["xyz", "live", "strong"];
        $this->revTest(inArray($a1, $a2), ["live", "strong"]);
        $a1 = ["live", "strong", "arp"];
        $this->revTest(inArray($a1, $a2), ["arp", "live", "strong"]);
    }

}